### PR TITLE
Cleanup of complete blob upload jobs

### PR DESCRIPTION
**Pull request by munishchouhan**
This PR will add job cleanup functionality

### CODE REVIEW

**Comment by munishchouhan** [comment]
tested in local:

**client:**

```
munish.chouhan@Munishs-MacBook-Pro blob_cache_testing % wave -i cr.seqera.io/public/nf-jdk:corretto-17-al2023 --wave-endpoint http://localhost:9090
ffb6c929b79f.ngrok.app/wt/bdda8ceef206/public/nf-jdk:corretto-17-al2023
munish.chouhan@Munishs-MacBook-Pro blob_cache_testing % docker pull ffb6c929b79f.ngrok.app/wt/bdda8ceef206/public/nf-jdk:corretto-17-al2023
corretto-17-al2023: Pulling from wt/bdda8ceef206/public/nf-jdk
6f59c9417334: Pull complete
7faae36eec06: Pull complete
81e8630da5f3: Pull complete
93fe533f05fe: Pull complete
Digest: sha256:cebacc8e86fed9625ac663f3d3eb1be39496185c611a66f735c81fec1aec9d02
Status: Downloaded newer image for ffb6c929b79f.ngrok.app/wt/bdda8ceef206/public/nf-jdk:corretto-17-al2023
ffb6c929b79f.ngrok.app/wt/bdda8ceef206/public/nf-jdk:corretto-17-al2023
```

**k8s:**
```

munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
No resources found in wave-local namespace.
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-3eb557e8cb2c42c4   0/1           4s         4s
transfer-de20239dc00daf13   0/1           4s         4s
transfer-c22f93d9488f9137   0/1           4s         4s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-de20239dc00daf13   0/1           16s        16s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
No resources found in wave-local namespace.
```

File: src/main/groovy/io/seqera/wave/service/counter/impl/RedisCounterProvider.groovy [hunk 1]
```diff
@@ -55,7 +59,7 @@ class RedisCounterProvider implements CounterProvider {
     @Override
     Map<String, Long> getAllMatchingEntries(String key, String pattern) {
         try(Jedis conn=pool.getResource() ) {
-            final scanResult = conn.hscan(key, "0", new ScanParams().match(pattern))
+            final scanResult = conn.hscan(key, "0", new ScanParams().match(pattern).count(hscanCount))
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/service/counter/impl/RedisCounterProvider.groovy:62]: This should not be here

**Comment by munishchouhan** [comment]
Its working now, I have added background propagation policy, which forces k8s GC to delete all the resources belongs to the deleted job

File: src/main/groovy/io/seqera/wave/service/counter/impl/RedisCounterProvider.groovy (hunk 1 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/counter/impl/RedisCounterProvider.groovy]: fixed

**Comment by munishchouhan** [comment]
Tested in dev

**client:**
```
munish.chouhan@Munishs-MacBook-Pro blob_cache_testing % wave -i cr.seqera.io/public/nf-jdk:corretto-20.0.2 --wave-endpoint https://wave.dev-tower.net
wave.dev-tower.net/wt/b72bbd54f82d/public/nf-jdk:corretto-20.0.2
munish.chouhan@Munishs-MacBook-Pro blob_cache_testing % docker pull wave.dev-tower.net/wt/b72bbd54f82d/public/nf-jdk:corretto-20.0.2
corretto-20.0.2: Pulling from wt/b72bbd54f82d/public/nf-jdk
8be3d01330d7: Pull complete
c1d57871a336: Pull complete
736942fcbd5a: Pull complete
6951eeac09c9: Pull complete
Digest: sha256:21cda30b4a887734958005d9caccbbfe2ec0c058195dde241e1e05e3ac5bc028
Status: Downloaded newer image for wave.dev-tower.net/wt/b72bbd54f82d/public/nf-jdk:corretto-20.0.2
wave.dev-tower.net/wt/b72bbd54f82d/public/nf-jdk:corretto-20.0.2
```
**k8s:**
```
munish.chouhan@Munishs-MacBook-Pro wave-dev % kubectl get pods -n wave-build
NAME                              READY   STATUS    RESTARTS   AGE
transfer-0a52dc48ea947341-8j7mc   1/1     Running   0          2s
transfer-2f047709becebfb4-tk7v8   1/1     Running   0          2s
transfer-7faa229081cf1fe9-lvfh5   1/1     Running   0          2s
munish.chouhan@Munishs-MacBook-Pro wave-dev % kubectl get jobs -n wave-build
NAME                        COMPLETIONS   DURATION   AGE
transfer-0a52dc48ea947341   0/1           5s         5s
transfer-2f047709becebfb4   0/1           5s         5s
transfer-7faa229081cf1fe9   0/1           5s         5s
munish.chouhan@Munishs-MacBook-Pro wave-dev % kubectl get jobs -n wave-build
No resources found in wave-build namespace.
munish.chouhan@Munishs-MacBook-Pro wave-dev % kubectl get pods -n wave-build
No resources found in wave-build namespace.
```

File: src/main/groovy/io/seqera/wave/service/blob/impl/DockerTransferStrategy.groovy [hunk 2]
```diff
@@ -73,4 +75,21 @@ class DockerTransferStrategy implements TransferStrategy {
...
+    protected ProcessBuilder deleteProcess(BlobCacheInfo info) {
+        final cli = new ArrayList<String>(10)
+        cli.add('docker')
+        cli.add('rm')
+        cli.add('-f')
+        cli.add(getName(info))
+        final builder = new ProcessBuilder()
+        builder.command(cli)
+        builder.redirectErrorStream(true)
+        return builder
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/service/blob/impl/DockerTransferStrategy.groovy:93]: It can ignored for docker, better to keep it simple

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/blob/impl/DockerTransferStrategy.groovy]: done

File: src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy [hunk 3]
```diff
@@ -82,6 +83,9 @@ class BlobCacheServiceImpl implements BlobCacheService {
     @Inject
     private HttpClientConfig httpConfig
 
+    @Inject
+    private CleanupStrategy cleanup
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy:87]: I think this should not be exposed as a service method and only kept in the K8s implementtion. Thoughts?

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy:87]: sorry, I dont understand
CleanupStrategy only provides information about whether cleanup is set for onsuccess, true or false. 
you want to remove CleanupStrategy?

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy:87]: My point is that the job can be deleted within thr

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy:87]: ok noted

**Comment by munishchouhan** [comment]
Tested new changes on local

**Client:**
```
% wave -i cr.seqera.io/public/nf-jdk:corretto-21.0.0-jemalloc --wave-endpoint http://localhost:9090
4ea8acee27a3.ngrok.app/wt/1964c87c7f1a/public/nf-jdk:corretto-21.0.0-jemalloc
munish.chouhan@Munishs-MacBook-Pro ~ % docker pull 4ea8acee27a3.ngrok.app/wt/1964c87c7f1a/public/nf-jdk:corretto-21.0.0-jemalloc
corretto-21.0.0-jemalloc: Pulling from wt/1964c87c7f1a/public/nf-jdk
9aa850931a9d: Already exists
09f0724b3f66: Pull complete
e889a8c4169b: Pull complete
321408116d8b: Pull complete
b70004fc5277: Pull complete
Digest: sha256:7698108095fcc286e7519462148e5b10a2dc18f3ec34aebb10b36af21274a788
Status: Downloaded newer image for 4ea8acee27a3.ngrok.app/wt/1964c87c7f1a/public/nf-jdk:corretto-21.0.0-jemalloc
4ea8acee27a3.ngrok.app/wt/1964c87c7f1a/public/nf-jdk:corretto-21.0.0-jemalloc
```

**k8s:**
```
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-56c5fc110314102a   0/1           4s         4s
transfer-61b3dfd47b162be5   0/1           4s         4s
transfer-6077cabdf90bdabc   0/1           2s         2s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                              READY   STATUS    RESTARTS   AGE
transfer-61b3dfd47b162be5-fl4mh   1/1     Running   0          6s
transfer-56c5fc110314102a-8cn5h   1/1     Running   0          6s
transfer-6077cabdf90bdabc-jfjkr   1/1     Running   0          4s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-61b3dfd47b162be5   0/1           18s        18s
transfer-6077cabdf90bdabc   0/1           16s        16s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                              READY   STATUS    RESTARTS   AGE
transfer-61b3dfd47b162be5-fl4mh   1/1     Running   0          20s
transfer-6077cabdf90bdabc-jfjkr   1/1     Running   0          18s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-61b3dfd47b162be5   0/1           29s        29s
transfer-6077cabdf90bdabc   0/1           27s        27s
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
No resources found in wave-local namespace.
munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
No resources found in wave-local namespace.
```

File: src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy (hunk 3 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy]: done

File: src/main/groovy/io/seqera/wave/service/blob/impl/KubeTransferStrategy.groovy [hunk 4]
```diff
@@ -69,9 +83,17 @@ class KubeTransferStrategy implements TransferStrategy {
         final pod = k8sService.getPod(podName)
         final terminated = k8sService.waitPod(pod, name, blobConfig.transferTimeout.toMillis())
         final stdout = k8sService.logsPod(podName, name)
-        return terminated
+
+        def result = terminated
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/service/blob/impl/KubeTransferStrategy.groovy:87]:
```suggestion
        final result = terminated
```

**Review by pditommaso — APPROVED** [review]

**Pull request merged**